x-pack/plugins/lens/public/pie_visualization/visualization.tsx

```diff
@@ -271,33 +252,53 @@ export const getPieVisualization = ({
     if (state?.layers.length === 0 || !frame.activeData) {
       return;
     }
-
-    const metricColumnsWithArrayValues = [];
+    const warningMessages = [];
 
     for (const layer of state.layers) {
       const { layerId, metric } = layer;
-      const rows = frame.activeData[layerId] && frame.activeData[layerId].rows;
+      const rows = frame.activeData[layerId]?.rows;
+      const numericColumn = frame.activeData[layerId]?.columns.find(
+        ({ meta }) => meta?.type === 'number'
+      );
+
       if (!rows || !metric) {
         break;
       }
-      const columnToLabel = frame.datasourceLayers[layerId].getOperationForColumnId(metric)?.label;
 
+      if (
+        numericColumn &&
+        state.shape === 'waffle' &&
+        checkTableForContainsSmallValues(frame.activeData[layerId], numericColumn.id, 1)
+      ) {
+        warningMessages.push(
+          <FormattedMessage
+            id="xpack.lens.pie.smallValuesWarningMessage"
+            defaultMessage="The received data contains too small values that can be incorrectly visualized in the {shape} chart"
```

Comment:
@KOTungseth maybe you could help out with the copy for this message?[ The screenshot of how it's used is here](https://user-images.githubusercontent.com/20072247/143047203-0fbc5d73-bcd5-4b50-87bf-b56f5426251d.png)

Reply:
How about "The data contains small category values that could be incorrectly represented by the waffle chart. Try switching to table or treemap to better represent these values."